### ISSUE TITLE
The same error when I'm trying to run model DANN and MFSAN

### ISSUE THREAD

**Issue by Wenbo-Qin** [Feaxure-fresh/TL-Fault-Diagnosis-Library#16]
When I put code python train.py --model_name MFSAN --source CWRU_0 --target CWRU_1 --train_mode single_source --cuda_device 0 to my termination and start to run it, it says that
 "TypeError: Trainer.__init__() got an unexpected keyword argument 'device'."

So I add the code from former
>    def __init__(self, args):
to
>    def __init__(self, args, device):
             self.device = device

It raises an other error after some epochs：
OSError: [Errno 24] Too many open files
I try to add codes 
import torch.multiprocessing
torch.multiprocessing.set_sharing_strategy('file_system')
but it still exists and I don't know how to slove it.

**Comment by Feaxure-fresh** [comment]
The code itself should not cause the error "TypeError: Trainer.init() got an unexpected keyword argument 'device'." You can restore the code to its original state, without modifying the `init` method or any other content. Then try running the command: 

```
python train.py --model_name MFSAN --source CWRU_0 --target CWRU_1 --train_mode single_source
```

If you still encounter similar errors, please provide more information, including the exact location of the error.

**Comment by Wenbo-Qin** [comment]
Hi, I've restored the code and try to run it, but there still raises an error. Here's the detail:
>Traceback (most recent call last):
  File "/Users/mon/JupyterProjects/UMDA/pythonProject/TL-Fault-Diagnosis-Library-main/train.py", line 143, in <module>
    trainer = importlib.import_module(f"models.{args.model_name}").Trainer(args)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mon/JupyterProjects/UMDA/pythonProject/TL-Fault-Diagnosis-Library-main/models/DANN.py", line 22, in __init__
    backbone=args.backbone, dropout=args.dropout).to(self.device)
                                                  ^^^^^^^^^^^^^^^
  File "/Users/mon/anaconda3/envs/tll/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1152, in to
    return self._apply(convert)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/mon/anaconda3/envs/tll/lib/python3.11/site-packages/torch/nn/modules/module.py", line 802, in _apply
    module._apply(fn)
  File "/Users/mon/anaconda3/envs/tll/lib/python3.11/site-packages/torch/nn/modules/module.py", line 802, in _apply
    module._apply(fn)
  File "/Users/mon/anaconda3/envs/tll/lib/python3.11/site-packages/torch/nn/modules/module.py", line 802, in _apply
    module._apply(fn)
  [Previous line repeated 2 more times]
  File "/Users/mon/anaconda3/envs/tll/lib/python3.11/site-packages/torch/nn/modules/module.py", line 825, in _apply
    param_applied = fn(param)
                    ^^^^^^^^^
  File "/Users/mon/anaconda3/envs/tll/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1150, in convert
    return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mon/anaconda3/envs/tll/lib/python3.11/site-packages/torch/cuda/__init__.py", line 293, in _lazy_init
    raise AssertionError("Torch not compiled with CUDA enabled")
AssertionError: Torch not compiled with CUDA enabled

My macOS version is 14.4.

Also, I use another PC without Nvidia (CPU only), run the same code in the command but got another error:

>(tll) C:\Users\Dell\Desktop\pythonProject\experiment\TL-Fault-Diagnosis-Library>python train.py --model_name DANN --source CWRU_0,CWRU_1 --target CWRU_2 --train_mode multi_source
Traceback (most recent call last):
  File "train.py", line 107, in <module>
    logger, args = creat_file(args)
  File "train.py", line 43, in creat_file
    logger = setlogger(args.save_path  + '.log')
  File "train.py", line 23, in setlogger
    fileHandler = logging.FileHandler(path)
  File "E:\Anaconda3\envs\tll\lib\logging\__init__.py", line 1147, in __init__
    StreamHandler.__init__(self, self._open())
  File "E:\Anaconda3\envs\tll\lib\logging\__init__.py", line 1176, in _open
    return open(self.baseFilename, self.mode, encoding=self.encoding)
OSError: [Errno 22] Invalid argument: 'C:\\Users\\Dell\\Desktop\\pythonProject\\experiment\\TL-Fault-Diagnosis-Library\\ckpt\\DANN\\multi_source\\[CWRU_0|CWRU_1]To[CWRU_2]_0730-094114.log'

However, I can run the code
>python train.py --model_name DAN --source CWRU_0 --target CWRU_1 --train_mode single_source --cuda_device 0

correctly.

**Comment by Feaxure-fresh** [comment]
Here is a more fluent version of the content:

On macOS, the error "AssertionError: Torch not compiled with CUDA enabled" indicates that PyTorch was not compiled with CUDA support. This issue is not due to the code itself. You should check if your device supports CUDA with PyTorch or switch to CPU-based training by setting `--cuda_device ''`.

In Windows, the error relates to something interesting. I haven't tested the latest version of the code on Windows. I discovered that the pipe symbol ("|") should not be used in Windows filenames, although it is acceptable in Ubuntu. To address this, I modified line-35 in `train.py` to avoid using "|" as a delimiter. This change makes filename separation less clear, and I am still searching for a more suitable delimiter.

**Comment by Feaxure-fresh** [comment]
In macOS, the error "AssertionError: Torch not compiled with CUDA enabled" suggests that your PyTorch installation isn't properly compiled with CUDA support. This issue doesn't stem from the code itself. To resolve this, verify whether your device supports running PyTorch with CUDA, or alternatively, you can train the model on the CPU by setting `--cuda_device ''`.

In Windows, this error is interesting. I haven't tested the latest version of the code on Windows. I discovered that the symbol "|" should not be used in Windows filenames, although it's acceptable in Ubuntu. To address this, I've modified line-35 of `train.py` to exclude the symbol "|" as a delimiter. This change makes filename separation less clear, and I am currently searching for a more suitable alternative delimiter.

**Comment by Wenbo-Qin** [comment]
Thanks for your reply, now I can run the code successfully in my Mac. But I still wonder that why code

```python
python train.py --model_name DAN --source CWRU_0 --target CWRU_1 --train_mode single_source --cuda_device 0
````
worked in the last 3 days but fails today in my Mac. That's strange 🤔.

**Comment by Wenbo-Qin** [comment]
Hi,
I modified the code In `DANN.py` in Windows by using

```python
    def __init__(self, args):
        if args.cuda_device and args.cuda_device.strip():
            self.device = torch.device("cuda:" + args.cuda_device)
        else:
            self.device = torch.device("cpu")
```
and the model DANN can run successfully.

But the error in Mac still exists even I try to modified the number of max process.

**Comment by Feaxure-fresh** [comment]
I have not encountered the problems you mentioned in the Issue on Ubuntu OS. Could you please send your contact information to my email so that we can discuss the details of your issue further?

**Comment by AI-Explorer-Lab** [comment]
Hello, actually, everything went well after that, but I haven't figure out which process led to that error. Thank you for your time and waiting for so long.

By the way, can this code achieve the open set issue? That's is, the overall contains 5 fault types and each souce domain contains 3 of 5 and the target domain contains all types? If so, how can I modify the code to achieve it? Thank you!

**Comment by Feaxure-fresh** [comment]
Our code already supports open-set scenarios, so no modifications are necessary. Simply navigate to the [Selective Transfer](#selective-transfer) section and follow the instructions to create your own open-set transfer or even a universal transfer.

**Comment by AI-Explorer-Lab** [comment]
> Our code already supports open-set scenarios, so no modifications are necessary. Simply navigate to the [Selective Transfer](#selective-transfer) section and follow the instructions to create your own open-set transfer or even a universal transfer.

Thank you. But when I try to run the code, it raises an error. The code is
>python train.py --model_name UDA --source CWRU_0-01,CWRU_1-01,CWRU_2-01 --target CWRU_3-012 --train_mode source_combine --cuda_device 0

and the error is 
> train.py", line 70, in <module>
    tgt, condition = args.target.split('_')[0], int(args.target.split('_')[1])
ValueError: invalid literal for int() with base 10: '3-012'

I think there's something wrong when use 'int(args.target.split('_')[1]', beacuse args.target.split('_')[1] will be recognized as 3-012, which is not an illegal number.

**Comment by AI-Explorer-Lab** [comment]
Finally, I run it successfully by using 'git clone' in command rather than download the ZIP file. That's strange!!! Thank you very much!!

**Comment by Feaxure-fresh** [comment]
I think this is likely because you ran the code before updating it to the latest version. :-P

**Comment by AI-Explorer-Lab** [comment]
Thank you very much, these codes do help!!
